soc/arm/nxp_imx/rt/lpm_rt1064.c

```diff
@@ -0,0 +1,458 @@
+/*
+ * Copyright (c) 2021, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Note: this file is linked to RAM. Any functions called while changing clocks
+ * to the flexspi modules must be linked to RAM, or within this file
+ */
+
+#include <init.h>
+#include <power_rt10xx.h>
+#include <zephyr.h>
+#include "clock_config.h"
+
+
+/*
+ * Clock configuration structures populated at boot time. These structures are
+ * used to reinitialize the PLLs after exiting low power mode.
+ */
+#ifdef CONFIG_INIT_ARM_PLL
+static clock_arm_pll_config_t arm_pll_config;
+#endif
+#ifdef CONFIG_INIT_VIDEO_PLL
+static clock_video_pll_config_t video_pll_config;
+#endif
+#ifdef CONFIG_INIT_ENET_PLL
+static clock_enet_pll_config_t enet_pll_config;
+#endif
+static clock_sys_pll_config_t sys_pll_config;
+static clock_usb_pll_config_t usb1_pll_config;
+
+#define IMX_RT_SYS_PFD_FRAC(reg, pfd_num) \
+	(((reg) >> (8U * pfd_num)) &\
+	CCM_ANALOG_PFD_528_PFD0_FRAC_MASK)
+#define IMX_RT_USB1_PFD_FRAC(reg, pfd_num) \
+	(((reg) >> (8U * pfd_num)) &\
+	CCM_ANALOG_PFD_480_PFD0_FRAC_MASK)
+
+uint8_t sys_pll_pfd0_frac;
+uint8_t sys_pll_pfd1_frac;
+uint8_t sys_pll_pfd2_frac;
+uint8_t sys_pll_pfd3_frac;
+
+uint8_t usb1_pll_pfd1_frac;
+uint8_t usb1_pll_pfd2_frac;
+uint8_t usb1_pll_pfd3_frac;
+
+uint32_t flexspi_div;
+
+/* Internal version of this function used, since we relocate this code to ITCM */
+static inline void CLOCK_SETMUX(clock_mux_t mux, uint32_t value)
+{
+	uint32_t busyShift;
+
+	busyShift = (uint32_t)CCM_TUPLE_BUSY_SHIFT(mux);
+	CCM_TUPLE_REG(CCM, mux) = (CCM_TUPLE_REG(CCM, mux) & (~CCM_TUPLE_MASK(mux))) |
+		  (((uint32_t)((value) << CCM_TUPLE_SHIFT(mux))) & CCM_TUPLE_MASK(mux));
+
+	assert(busyShift <= CCM_NO_BUSY_WAIT);
+
+	/* Clock switch need Handshake? */
+	if (busyShift != CCM_NO_BUSY_WAIT) {
+		/* Wait until CCM internal handshake finish. */
+		while ((CCM->CDHIPR & ((1UL << busyShift))) != 0UL) {
+		}
+	}
+}
```

Comment:
I'll update this function to use correct coding guidelines. However, I believe it should remain in this file, because this function needs to be linked into ITCM (since it is invoked to configure the clocks of the flash memory and SRAM).

The function could be moved to another file that is also relocated to ITCM, but I would argue it is best to keep all code touching the flash and SRAM clocks in one place. Bugs involving reading from the flash while changing its clocks often are very hard to narrow down, and adding something like a `LOG_DBG` call in this function would be enough to cause issues. Keeping all flash clocking code in one file that is moved to ITCM reduces the risk of a user unintentionally adding function calls here that could introduce bugs.